File: July2021/IsomorphicStrings.java
class Solution {
  // TC : O(n)
  public boolean isIsomorphic(String s, String t) {
    if(s.length()!=t.length()){
      return false;
    }
    Map<Character, Character> sTMap = new HashMap<>();

    Map<Character, Character> tSMap = new HashMap<>();

    for(int i=0;i<s.length();i++){
      Character sChar = s.charAt(i);
      Character tChar = t.charAt(i);

      if(sTMap.containsKey(sChar)){
        if(sTMap.get(sChar)!=tChar)
          return false;
      }
      if(tSMap.containsKey(tChar)){
        if(tSMap.get(tChar)!=sChar)
          return false;
      }

      sTMap.put(sChar, tChar);
      tSMap.put(tChar, sChar);
    }

    return true;
  }
}

File: July2021/C++/IsomorphicStrings.cpp
class Solution
{
public:
    bool isIsomorphic(string s, string t)
    {
        if (s.size() != t.size())
            return false;
        map<char, char> m;
        for (int i = 0; i < s.size(); i++)
        {
            if (m.find(s[i]) == m.end())
            {
                m[s[i]] = t[i];
            }
            else
            {
                if (m[s[i]] != t[i])
                    return false;
            }
        }
        m.clear();
        for (int i = 0; i < s.size(); i++)
        {
            if (m.find(t[i]) == m.end())
                m[t[i]] = s[i];
            else
            {
                if (m[t[i]] != s[i])
                    return false;
            }
        }
        return true;
    }
};